pkg/process/monitor/process_monitor.go

```diff
@@ -242,159 +214,220 @@ func (pm *ProcessMonitor) Initialize() error {
+	if _, exists := pm.processExitCallbacks.Load(callback); exists {
+		return nil, errors.New("same callback can't be registered twice")
 	}
 
 	// check if the sibling Exec callback exist
-	if callback.Event == EXIT && callback.Metadata != ANY {
+	if callback.FilterType != ANY {
 		foundSibling := false
-		for _, c := range pm.procEventCallbacks[EXEC] {
-			if c.Metadata == callback.Metadata && c.Regex.String() == callback.Regex.String() {
+		pm.processExecCallbacks.Range(func(key, _ any) bool {
+			execCallback := key.(*ProcessCallback)
+			if callback.FilterType == execCallback.FilterType && callback.Regex.String() == execCallback.Regex.String() {
 				foundSibling = true
-				break
+				// stopping iteration.
+				return false
 			}
-		}
+			return true
+		})
 		if !foundSibling {
 			return nil, errors.New("no Exec callback has been found with the same Metadata and Regex, please Subscribe(Exec callback, Metadata) first")
 		}
 	}
 
-	pm.procEventCallbacks[callback.Event] = append(pm.procEventCallbacks[callback.Event], callback)
+	pm.processExitCallbacks.Store(callback, struct{}{})
 
 	// UnSubscribe()
 	return func() {
-		pm.m.Lock()
-		defer pm.m.Unlock()
-
-		// we are scanning all callbacks remove the one we registered
-		// and remove it from the pm.procEventCallbacks[callback.Event] list
-		for i, c := range pm.procEventCallbacks[callback.Event] {
-			if c == callback {
-				l := len(pm.procEventCallbacks[callback.Event])
-				pm.procEventCallbacks[callback.Event][i] = pm.procEventCallbacks[callback.Event][l-1]
-				pm.procEventCallbacks[callback.Event] = pm.procEventCallbacks[callback.Event][:l-1]
-				return
-			}
-		}
+		pm.processExitCallbacks.Delete(callback)
 	}, nil
 }
 
+// Stop decreasing the refcount, and if we reach 0 we terminate the main event loop.
 func (pm *ProcessMonitor) Stop() {
-	pm.m.Lock()
-	if pm.refcount == 0 {
-		pm.m.Unlock()
-		return
-	}
-
-	pm.refcount--
-	if pm.refcount > 0 {
-		pm.m.Unlock()
+	if pm.refcount.Dec() != 0 {
+		if pm.refcount.Load() < 0 {
+			pm.refcount.Swap(0)
+		}
 		return
 	}
 
-	pm.isInitialized = false
-	pm.m.Unlock()
 	close(pm.done)
+	// that's being done for testing purposes.
+	// As tests are running altogether, initOne and processMonitor are being created only once per compilation unit
+	// thus, the first test works without an issue, but the second test has troubles.
+	initOnce = sync.Once{}
+	pm.processExecCallbacks = sync.Map{}
+	pm.processExitCallbacks = sync.Map{}
 	pm.wg.Wait()
 }
 
-func (pm *ProcessMonitor) logStats() {
-	log.Debugf("process monitor stats - total events: %v; exec events: %v; exit events: %v", pm.eventCount, pm.execCount, pm.exitCount)
+// FindDeletedProcesses returns the terminated PIDs from the given map.
+func FindDeletedProcesses[V any](pids map[int32]V) map[int32]struct{} {
+	existingPids := make(map[int32]struct{}, len(pids))
+
+	procIter := func(pid int) error {
+		if _, exists := pids[int32(pid)]; exists {
+			existingPids[int32(pid)] = struct{}{}
```

Comment:
```
res := copy(pids)
....
	procIter := func(pid int) error {
		if _, exists := pids[int32(pid)]; exists {
          delete(res, pid)
        }
    }
   util.WithAllProcs(.....)
   return res
    
```

Reply:
copy works only on array, and unfortunately that's not the scenario here